### PR TITLE
[Rust] Early Stage Optimistic PR for Options Concept Exercise

### PR DESCRIPTION
**Pull request by seanchen1991**
This early stage optimistic PR for a Rust concept exercise on `Option`s contains the following file:
- instructions.md
- design.md
- working example
- tests

Files that still need to be added:
- introduction.md
- after.md
- hints.md

Otherwise, the core of the exercise is set and ready for review!

### CODE REVIEW

File: languages/rust/exercises/concept/options/tests/options.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    let level: u32 = 34;
+    let dead_player = Player {
+        health: 0,
+        mana: Some(0),
+        level,
+    };
+    let revived_player = dead_player.revive();
+    assert_eq!(revived_player.is_some(), true);
+
+    let revived_player = revived_player.unwrap();
```
Comment (coriolinus) [inline languages/rust/exercises/concept/options/tests/options.rs:14]:
```suggestion
    let revived_player = dead_player.revive().expect("reviving a dead player must return Some(player)");
```

Shorter and gives a clearer error message.

File: languages/rust/exercises/concept/options/tests/options.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+use options::*;
+
+#[test]
+fn test_reviving_dead_player() {
+    let level: u32 = 34;
+    let dead_player = Player {
+        health: 0,
+        mana: Some(0),
+        level,
```
Comment (coriolinus) [inline languages/rust/exercises/concept/options/tests/options.rs:9]:
```suggestion
        34,
```

Just use a constant here; you don't need to break out a separate `level` variable.

File: languages/rust/exercises/concept/options/tests/options.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    };
+    let revived_player = dead_player.revive();
+    assert_eq!(revived_player.is_some(), true);
+
+    let revived_player = revived_player.unwrap();
+    assert_eq!(revived_player.health, 100);
+    assert_eq!(revived_player.mana, Some(100));
+    assert_eq!(revived_player.level, level);
```
Comment (coriolinus) [inline languages/rust/exercises/concept/options/tests/options.rs:17]:
```suggestion
    assert_eq!(dead_player.level, revived_player.level);
```

File: languages/rust/exercises/concept/options/tests/options.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+#[test]
+#[ignore]
+fn test_reviving_alive_player() {
+    let level: u32 = 8;
+    let alive_player = Player {
+        health: 1,
+        mana: None,
+        level,
```
Comment (coriolinus) [inline languages/rust/exercises/concept/options/tests/options.rs:27]:
```suggestion
        8,
```

Again, no need to break out a `level` variable.

File: languages/rust/exercises/concept/options/tests/options.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    assert_eq!(accomplished_wizard.health, health);
+    assert_eq!(accomplished_wizard.mana, Some(mana - mana_cost));
+    assert_eq!(accomplished_wizard.level, level);
+}
+
+#[test]
+#[ignore]
+fn test_cast_spell_with_insufficient_mana() {
+    let health = 56;
+    let mana = 0;
```
Comment (coriolinus) [inline languages/rust/exercises/concept/options/tests/options.rs:56]:
```suggestion
    let mana = 2;
```

By using a non-0 value, we reduce the chance that a student's accidentally done the right thing by applying a default somewhere.

File: languages/rust/exercises/concept/options/.meta/example.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            Some(m) => {
+                if m >= mana_cost {
+                    self.m -= mana_cost;
+                    return mana_cost * 2;
+                } else {
+                    return 0;
+                }
+            }
+            None => 0,
+        }
```
[the remark this answers is not in the document]
Comment (seanchen1991) [reply languages/rust/exercises/concept/options/.meta/example.rs:34]: Something like this in the `None` case?

```rust
None => {
    self.health = match self.health.checked_sub(5) {
        Some(new_health) => new_health,
        None => 0,
    };
    0
} 
```

[the remark this answers is not in the document]
Comment (coriolinus) [reply languages/rust/exercises/concept/options/.meta/example.rs:34]: But it's a bit better as 

```rust
None => {
    self.health = match self.health.checked_sub(5).unwrap_or_default();
    0
} 
```

**Comment by efx** [comment]
@coriolinus @seanchen1991 y'all okay with me approving and merging? Core work looks solid and we can continue addressing the feedback in subsequent PRs. 🚀

**Comment by coriolinus** [comment]
Yes, conditional on breaking out the remaining feedback as issues. If we just leave it as comments on this PR, then it will inevitably get lost and forgotten.

**Pull request merged**